ldp/alg/algorithms.py

```diff
@@ -132,67 +133,111 @@ def gvizify(x: Any) -> str:
-            extract_answer_fn = cast(
-                Callable[[TData], Awaitable[TAnswer]], extract_answer_fn
-            )
-            answers = await asyncio.gather(*(extract_answer_fn(x) for x in sampled))
-        else:
-            extract_answer_fn = cast(Callable[[TData], TAnswer], extract_answer_fn)
-            answers = [extract_answer_fn(x) for x in sampled]
-
-        # Compute consensus: mode of the sampled answers
-        grouped_consensus[group_key] = collections.Counter(answers).most_common()
-        # NOTE: If there are multiple modes, just use the first one
-        consensus: TAnswer = grouped_consensus[group_key][0][0]
-        if ideal_answer_fn != NO_IDEAL_ANSWER_FN:
+    rand = random.Random(seed) if seed is not None else seed
+    grouped_consensus: dict[
+        TGroupKey, tuple[list[tuple[TAnswer, int]], float, float]
+    ] = {}
+
+    async def index(group_key: TGroupKey, group: list[TData]) -> int:
+        bins, consensus, mean, ste = await evaluate_consensus(
+            group, extract_answer_fn, num_samples, rand
+        )
+        grouped_consensus[group_key] = bins, mean, ste
+        if ideal_answer_fn != NO_IDEAL_ANSWER_FN:  # If we have an ideal
             # Assume all items in the group have the same ideal answer
-            ideal_count += consensus == ideal_answer_fn(group[0])
+            return consensus == ideal_answer_fn(group[0])
+        return 0
 
+    ideal_count = sum(await asyncio.gather(*itertools.starmap(index, groups.items())))
     return grouped_consensus, ideal_count / len(groups) if groups else 0.0
 
 
+async def evaluate_consensus(
+    data: Sequence[TData],
+    extract_answer_fn: Callable[[TData], TAnswer | Awaitable[TAnswer]],
+    num_samples: int,
+    seed: random.Random | int | None = None,
+) -> tuple[list[tuple[TAnswer, int]], TAnswer, float, float]:
+    """
+    Create consensus bins given data.
+
+    Args:
+        data: Data to evaluate consensus upon, length is called N.
+        extract_answer_fn: Function to extract the actual answer from a datum. It can
+            be async so this can be done using a LLM call.
+        num_samples: Number of samples to choose from the N total.
+        seed: Optional seed for sampling.
+
+    Returns:
+        Four-tuple of consensus bins generated by collections.Counter.most_common, the
+            consensus answer, and the mean and standard error of the consensus.
+    """
+    if isinstance(seed, int):
+        rand: ModuleType | random.Random = random.Random(seed)
+    elif isinstance(seed, random.Random):
+        rand = seed
+    else:
+        rand = random
+
+    if len(data) < num_samples:  # Too few items, sample with replacement
+        sampled = [rand.choice(data) for _ in range(num_samples)]
+    else:  # Sample without replacement
+        sampled = rand.sample(data, num_samples)
+
+    # Get answers for the sampled data
+    if is_coroutine_callable(extract_answer_fn):
```

Comment:
See note here: https://github.com/Future-House/ldp/pull/212/files#r1920598727

Reply:
Fixed this separately in https://github.com/Future-House/ldp/pull/213